core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java

```diff
@@ -0,0 +1,138 @@
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSpiBackwardCompatibility
+{
+    private static final Map<String, Set<String>> BACKWARD_INCOMPATIBLE_CHANGES = ImmutableMap.<String, Set<String>>builder()
+            // When updating this map, please try to remove backward incompatible changes for old versions.
+            // Also consider mentioning backward incompatible changes in release notes.
+            .put("375", ImmutableSet.of(
+                    "Constructor: public io.trino.spi.connector.TableProcedureExecutionMode(boolean,boolean)",
+                    "Constructor: public io.trino.spi.eventlistener.QueryStatistics(java.time.Duration,java.time.Duration,java.time.Duration,java.time.Duration,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,long,long,long,long,long,long,long,long,long,long,long,long,long,long,long,double,double,java.util.List<io.trino.spi.eventlistener.StageGcStatistics>,int,boolean,java.util.List<io.trino.spi.eventlistener.StageCpuDistribution>,java.util.List<java.lang.String>,java.util.Optional<java.lang.String>)"))
+            .buildOrThrow();
+
+    @Test
+    public void testSpiSingleVersionBackwardCompatibility()
+            throws Exception
+    {
+        Set<String> backwardIncompatibleChanges = BACKWARD_INCOMPATIBLE_CHANGES.getOrDefault(getPreviousSpiVersion(), ImmutableSet.of());
+        assertThat(getCurrentSpi()).containsAll(difference(getPreviousSpi(), backwardIncompatibleChanges));
+    }
+
+    private static String getPreviousSpiVersion()
+    {
+        try {
+            URL resource = TestSpiBackwardCompatibility.class.getClassLoader().getResource("io/trino/spi/previous-trino-spi-version.txt");
+            requireNonNull(resource, "version resource not found");
+            try (InputStream inputStream = resource.openStream();
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, UTF_8))) {
+                String spiVersion = reader.readLine();
+                if (spiVersion == null || spiVersion.isBlank() || reader.readLine() != null) {
+                    throw new IllegalStateException("Malformed version resource");
+                }
+                return spiVersion.strip();
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static Set<String> getCurrentSpi()
+            throws IOException
+    {
+        return getSpiEntities(getSystemClassLoader(), true);
+    }
+
+    private static Set<String> getPreviousSpi()
+            throws Exception
+    {
+        try (Stream<Path> list = Files.list(Path.of("target", "released-artifacts"))) {
```

Comment:
cwd may or may not be module's dir.

- does it work when run from IDE (with your default settings)?
- does it work when run as `mvn test -pl :trino-spi`?
- does it work when run as `cd core/trino-spi && mvn test`?

Reply:
Yes. Yes. Yes.

Reply:
Awesome! 

Curious, how does it work then? i remember we had some problem with similar code here
https://github.com/trinodb/trino/blob/7cac412b9e69acf39b2e7c2111767f16998619d7/testing/trino-benchto-benchmarks/src/test/java/io/trino/sql/planner/AbstractCostBasedPlanTest.java#L131-L137

Reply:
I recall that too. Maybe something changed in maven?